File: granola_share/summarize.py
"""Write lecture notes from the transcript with a local Ollama model.

Granola's own summary comes from whatever model Granola runs. This lets you pick
the model (usually a big local one) and write the notes from the raw transcript instead.
Transcripts longer than the model's context are summarized in parts, then merged.
"""

from __future__ import annotations

import re
from typing import Callable

import httpx

from .config import Config
from .granola import Meeting

CHARS_PER_TOKEN = 3.5  # rough, for English speech
# Below this (a couple of minutes of speech) there's too little to fill the notes: small models then
# pad and loop. Such lectures keep Granola's own summary.
MIN_TRANSCRIPT_CHARS = 1500
# Good notes are well under 2,000 tokens. Without a cap, a small model that starts repeating itself
# never stops: Ollama keeps shifting its context and generating (39,000 tokens seen, from llama3.2:3b).
MAX_NOTES_TOKENS = 4096

STRUCTURE = """Use exactly this structure, and skip any section the lecture has nothing for:

## Overview
Two to four sentences: what the lecture covered and how it fits the course.

## Key concepts
Bullets. Bold the term, then explain it in one or two sentences the way the lecturer did.

## Details and examples
Worked examples, derivations, formulas (LaTeX in $...$), code, and demonstrations, in the order they were taught.

## Announcements
Deadlines, exams, assignments, and readings, with dates exactly as said.

## Review questions
Three to five questions a student should be able to answer after this lecture."""

RULES = """Rules:
- Use only what is in the transcript. Never invent facts, dates, or examples.
- The transcript comes from speech recognition: fix obvious mis-hearings of technical terms, and skip filler, small talk, and audio problems.
- Write in the language of the lecture. Output Markdown only, with no preamble."""

# (cfg, model, prompt, num_ctx) -> text
ChatFn = Callable[[Config, str, str, int], str]


class RunawayOutput(RuntimeError):
    """The model didn't finish properly: it ran to the length cap, or repeated itself."""


def ollama_generate(cfg: Config, model: str, prompt: str, num_ctx: int) -> str:
    body = {
        "model": model,
        "messages": [{"role": "user", "content": prompt}],
        "stream": False,
        "think": False,
        # A mild repeat penalty: stronger ones mangle Markdown, whose bullets legitimately repeat.
        "options": {"temperature": 0.2, "num_ctx": num_ctx, "num_predict": min(MAX_NOTES_TOKENS, num_ctx // 2),
                    "repeat_penalty": 1.1, "repeat_last_n": 256},
    }
    # A long lecture on a big model, or a slow computer, takes minutes; fine, this runs in the background.
    # Output is capped, so this only has to cover reading the transcript plus 4,096 tokens.
    r = httpx.post(f"{cfg.ollama_host.rstrip('/')}/api/chat", json=body, timeout=httpx.Timeout(900, connect=10))
    r.raise_for_status()
    data = r.json()
    if data.get("done_reason") == "length":
        raise RunawayOutput(f"{model} kept writing past {body['options']['num_predict']} tokens without finishing "
                            "(small models sometimes loop), so Granola's notes stay")
    return data["message"]["content"]


def repetitive(text: str, times: int = 5) -> bool:
    """The same line over and over: a model stuck in a loop, not notes."""
    lines = [ln.strip() for ln in text.splitlines() if len(ln.strip()) >= 12]
    if not lines:
        return False
    counts: dict[str, int] = {}
    for ln in lines:
        counts[ln] = counts.get(ln, 0) + 1
    return max(counts.values()) >= times or (len(lines) >= 20 and len(counts) < len(lines) / 2)


def model_context(cfg: Config, model: str) -> int | None:
    """The model's native context length, from Ollama's /api/show."""
    try:
        r = httpx.post(f"{cfg.ollama_host.rstrip('/')}/api/show", json={"model": model}, timeout=10)
        r.raise_for_status()
        for key, value in (r.json().get("model_info") or {}).items():
            if key.endswith(".context_length"):
                return int(value)
    except Exception:
        pass
    return None


def context_size(cfg: Config, model: str, show=model_context) -> int:
    cap = max(4096, cfg.summary_max_context)
    native = show(cfg, model)
    return min(cap, native) if native else cap


def transcript_budget(ctx: int) -> int:
    """Characters of transcript that fit in one call, leaving room for instructions and the answer."""
    reserved = min(4096, ctx // 3)
    return int((ctx - reserved) * CHARS_PER_TOKEN)


def wants_summary(m: Meeting, cfg: Config) -> bool:
    return cfg.ollama_enabled and cfg.summary_enabled and len(m.transcript.strip()) >= MIN_TRANSCRIPT_CHARS


def clean_output(text: str) -> str:
    text = re.sub(r"<think>.*?</think>", "", text or "", flags=re.DOTALL).strip()
    fence = re.fullmatch(r"```(?:markdown|md)?\s*\n(.*)\n```", text, flags=re.DOTALL)
    text = (fence.group(1) if fence else text).strip()
    if repetitive(text):
        raise RunawayOutput("the model repeated itself instead of writing notes, so Granola's notes stay")
    return text


def _pieces(line: str, max_chars: int) -> list[str]:
    """Break one over-long line at sentence ends, then at spaces, then anywhere."""
    if len(line) <= max_chars:
        return [line]
    out, cur = [], ""
    for sentence in re.split(r"(?<=[.!?])\s+", line):
        while len(sentence) > max_chars:
            cut = sentence.rfind(" ", 0, max_chars)
            cut = cut if cut > max_chars // 2 else max_chars
            head, sentence = sentence[:cut], sentence[cut:].lstrip()
            if cur:
                out.append(cur)
                cur = ""
            out.append(head)
        if cur and len(cur) + 1 + len(sentence) > max_chars:
            out.append(cur)
            cur = ""
        cur = f"{cur} {sentence}" if cur else sentence
    if cur:
        out.append(cur)
    return out


def split_transcript(text: str, max_chars: int) -> list[str]:
    """Split on line breaks into parts of at most max_chars."""
    parts, cur, size = [], [], 0
    for raw in text.splitlines():
        for line in _pieces(raw, max_chars):
            if cur and size + len(line) + 1 > max_chars:
                parts.append("\n".join(cur))
                cur, size = [], 0
            cur.append(line)
            size += len(line) + 1
    if cur:
        parts.append("\n".join(cur))
    return [p for p in parts if p.strip()]


def _header(m: Meeting) -> str:
    lines = [f"Lecture: {m.title}", f"Date: {m.date[:10]}"]
    if m.folder:
        lines.append(f"Granola folder: {m.folder}")
    return "\n".join(lines)


def whole_prompt(m: Meeting, transcript: str) -> str:
    return (f"You are an expert note-taker for university lectures. Write study notes for this lecture "
            f"from its transcript.\n\n{STRUCTURE}\n\n{RULES}\n\n{_header(m)}\n\nTranscript:\n{transcript}")


def part_prompt(m: Meeting, part: str, i: int, n: int) -> str:
    return (f"You are taking notes on part {i} of {n} of a lecture transcript. Write detailed Markdown bullet "
            f"notes for this part only: every concept, definition, example, formula, and announcement, in order. "
            f"They will be merged with the other parts later, so write no introduction or conclusion.\n\n"
            f"{RULES}\n\n{_header(m)}\n\nTranscript part {i} of {n}:\n{part}")


def condense_prompt(m: Meeting, notes: list[str]) -> str:
    joined = "\n\n".join(f"### Notes {i}\n{n}" for i, n in enumerate(notes, 1))
    return (f"Combine these notes on consecutive parts of one lecture into one set of detailed Markdown bullet "
            f"notes. Remove repetition but keep every distinct fact.\n\n{RULES}\n\n{_header(m)}\n\n{joined}")


def merge_prompt(m: Meeting, notes: list[str]) -> str:
    joined = "\n\n".join(f"### Part {i}\n{n}" for i, n in enumerate(notes, 1))
    return (f"You are an expert note-taker for university lectures. Below are notes on consecutive parts of one "
            f"lecture. Merge them into one set of study notes, removing repetition and keeping every distinct "
            f"fact.\n\n{STRUCTURE}\n\n{RULES}\n\n{_header(m)}\n\n{joined}")


def summarize_transcript(m: Meeting, cfg: Config, chat: ChatFn | None = None, show=None) -> str:
    """Our own study notes for a lecture, written from its transcript."""
    chat = chat or ollama_generate
    model = cfg.effective_summary_model
    ctx = context_size(cfg, model, show or model_context)
    budget = transcript_budget(ctx)
    text = m.transcript.strip()
    if len(text) <= budget:
        return clean_output(chat(cfg, model, whole_prompt(m, text), ctx))
    parts = split_transcript(text, budget)
    notes = [clean_output(chat(cfg, model, part_prompt(m, p, i, len(parts)), ctx)) for i, p in enumerate(parts, 1)]
    # Very long lectures: fold pairs of part-notes together until the merge fits in one call.
    while len(notes) > 1 and sum(len(n) for n in notes) > budget:
        notes = [clean_output(chat(cfg, model, condense_prompt(m, notes[i:i + 2]), ctx)) if i + 1 < len(notes)
                 else notes[i] for i in range(0, len(notes), 2)]
    return clean_output(chat(cfg, model, merge_prompt(m, notes), ctx))

File: engine/src/StudyStash.Core/Summarize.cs
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Text.RegularExpressions;

namespace StudyStash.Core;

/// <summary>(cfg, model, prompt, num_ctx) → the model's answer.</summary>
public delegate Task<string> ChatFn(Config cfg, string model, string prompt, int numCtx);

/// <summary>(cfg, model) → the model's own context length, or null when Ollama doesn't say.</summary>
public delegate Task<int?> ShowFn(Config cfg, string model);

/// <summary>The model didn't finish properly: it ran to the length cap, or repeated itself.</summary>
public sealed class RunawayOutputException(string message) : Exception(message);

/// <summary>
/// Write lecture notes from the transcript with a local Ollama model. Granola's own summary comes from whatever
/// model Granola runs; this writes the notes with the model you pick, from the raw transcript. Transcripts longer
/// than the model's context are summarized in parts, then merged.
/// </summary>
public static partial class Summarize
{
    public const double CharsPerToken = 3.5; // rough, for English speech
    // Below this (a couple of minutes of speech) there's too little to fill the notes: small models then pad and
    // loop. Such lectures keep Granola's own summary.
    public const int MinTranscriptChars = 1500;
    // Good notes are well under 2,000 tokens. Without a cap, a small model that starts repeating itself never
    // stops: Ollama keeps shifting its context and generating (39,000 tokens seen, from llama3.2:3b).
    public const int MaxNotesTokens = 4096;

    // .ReplaceLineEndings: on Windows, git checks this file out with "\r\n", which would change the prompts.
    public static readonly string Structure = """
        Use exactly this structure, and skip any section the lecture has nothing for:

        ## Summary
        Two to four sentences: what the lecture covered and how it fits the course.

        ## Key points
        Bullets: the ideas to remember, each in one or two sentences, the way the lecturer explained them.

        ## Definitions
        Bullets: the bold term, a colon, then what it means in one sentence.

        ## Details and examples
        Worked examples, derivations, formulas (LaTeX in $...$), code, and demonstrations, in the order they were taught.

        ## Announcements
        Deadlines, exams, assignments, and readings, with dates exactly as said.

        ## Questions to review
        Three to five numbered questions a student should be able to answer after this lecture.
        """.ReplaceLineEndings("\n");

    public static readonly string Rules = """
        Rules:
        - Use only what is in the transcript. Never invent facts, dates, or examples.
        - The transcript comes from speech recognition: fix obvious mis-hearings of technical terms, and skip filler, small talk, and audio problems.
        - Write in the language of the lecture. Output Markdown only, with no preamble.
        """.ReplaceLineEndings("\n");

    public static async Task<string> OllamaGenerateAsync(Config cfg, string model, string prompt, int numCtx,
        HttpClient? http = null)
    {
        int numPredict = Math.Min(MaxNotesTokens, numCtx / 2);
        var body = new JsonObject
        {
            ["model"] = model,
            ["messages"] = new JsonArray(new JsonObject { ["role"] = "user", ["content"] = prompt }),
            ["stream"] = false,
            ["think"] = false,
            // A mild repeat penalty: stronger ones mangle Markdown, whose bullets legitimately repeat.
            ["options"] = new JsonObject
            {
                ["temperature"] = 0.2, ["num_ctx"] = numCtx, ["num_predict"] = numPredict,
                ["repeat_penalty"] = 1.1, ["repeat_last_n"] = 256,
            },
        };
        // A long lecture on a big model, or a slow computer, takes minutes; fine, this runs in the background.
        // Output is capped, so this only has to cover reading the transcript plus 4,096 tokens.
        var data = await Ollama.PostAsync(cfg.OllamaHost, "/api/chat", body, TimeSpan.FromSeconds(900), http);
        if (Py.AsString(data?["done_reason"]) == "length")
            throw new RunawayOutputException($"{model} kept writing past {numPredict} tokens without finishing "
                + "(small models sometimes loop), so Granola's notes stay");
        return Py.AsString(data?["message"]?["content"]) ?? throw new InvalidDataException("Ollama sent no message");
    }

    /// <summary>The same line over and over: a model stuck in a loop, not notes.</summary>
    public static bool Repetitive(string text, int times = 5)
    {
        var lines = Py.SplitLines(text).Select(Py.Strip).Where(l => l.Length >= 12).ToList();
        if (lines.Count == 0) return false;
        var counts = lines.GroupBy(l => l, StringComparer.Ordinal).ToDictionary(g => g.Key, g => g.Count());
        return counts.Values.Max() >= times || (lines.Count >= 20 && counts.Count < lines.Count / 2.0);
    }

    /// <summary>The model's own context length, from Ollama's /api/show.</summary>
    public static async Task<int?> ModelContextAsync(Config cfg, string model)
    {
        try
        {
            var data = await Ollama.PostAsync(cfg.OllamaHost, "/api/show", new JsonObject { ["model"] = model },
                TimeSpan.FromSeconds(10));
            if (data?["model_info"] is JsonObject info)
                foreach (var (key, value) in info)
                    if (key.EndsWith(".context_length", StringComparison.Ordinal) && value is JsonValue v)
                        return v.GetValueKind() == JsonValueKind.Number ? (int)v.GetValue<double>() : int.Parse(Py.Str(v));
        }
        catch (Exception)
        {
            // Ollama not running, an old version, or a model it doesn't know: use our own cap.
        }
        return null;
    }

    public static async Task<int> ContextSizeAsync(Config cfg, string model, ShowFn? show = null)
    {
        int cap = Math.Max(4096, cfg.SummaryMaxContext);
        int? native = await (show ?? ModelContextAsync)(cfg, model);
        return native is int n && n != 0 ? Math.Min(cap, n) : cap;
    }

    /// <summary>Characters of transcript that fit in one call, leaving room for instructions and the answer.</summary>
    public static int TranscriptBudget(int ctx)
    {
        int reserved = Math.Min(4096, ctx / 3);
        return (int)((ctx - reserved) * CharsPerToken);
    }

    public static bool WantsSummary(Meeting m, Config cfg) =>
        cfg.OllamaEnabled && cfg.SummaryEnabled && Py.Strip(m.Transcript).Length >= MinTranscriptChars;

    [GeneratedRegex("<think>.*?</think>", RegexOptions.Singleline)]
    private static partial Regex Thinking();

    [GeneratedRegex(@"\A```(?:markdown|md)?\s*\n(.*)\n```\z", RegexOptions.Singleline)]
    private static partial Regex Fenced();

    [GeneratedRegex(@"(?<=[.!?])\s+")]
    private static partial Regex SentenceEnd();

    public static string CleanOutput(string? text)
    {
        string t = Py.Strip(Thinking().Replace(text ?? "", ""));
        var fence = Fenced().Match(t);
        t = Py.Strip(fence.Success ? fence.Groups[1].Value : t);
        if (Repetitive(t))
            throw new RunawayOutputException("the model repeated itself instead of writing notes, so Granola's notes stay");
        return t;
    }

    /// <summary>Break one over-long line at sentence ends, then at spaces, then anywhere.</summary>
    internal static List<string> Pieces(string line, int maxChars)
    {
        if (line.Length <= maxChars) return [line];
        var output = new List<string>();
        string cur = "";
        foreach (string s in SentenceEnd().Split(line))
        {
            string sentence = s;
            while (sentence.Length > maxChars)
            {
                int cut = sentence.LastIndexOf(' ', maxChars - 1);
                cut = cut > maxChars / 2 ? cut : maxChars;
                string head = sentence[..cut];
                sentence = Py.LStrip(sentence[cut..]);
                if (cur.Length > 0)
                {
                    output.Add(cur);
                    cur = "";
                }
                output.Add(head);
            }
            if (cur.Length > 0 && cur.Length + 1 + sentence.Length > maxChars)
            {
                output.Add(cur);
                cur = "";
            }
            cur = cur.Length > 0 ? $"{cur} {sentence}" : sentence;
        }
        if (cur.Length > 0) output.Add(cur);
        return output;
    }

    /// <summary>Split on line breaks into parts of at most maxChars.</summary>
    public static List<string> SplitTranscript(string text, int maxChars)
    {
        var parts = new List<string>();
        var cur = new List<string>();
        int size = 0;
        foreach (string raw in Py.SplitLines(text))
        {
            foreach (string line in Pieces(raw, maxChars))
            {
                if (cur.Count > 0 && size + line.Length + 1 > maxChars)
                {
                    parts.Add(string.Join("\n", cur));
                    cur = [];
                    size = 0;
                }
                cur.Add(line);
                size += line.Length + 1;
            }
        }
        if (cur.Count > 0) parts.Add(string.Join("\n", cur));
        return parts.Where(p => Py.Strip(p).Length > 0).ToList();
    }

    static string Header(Meeting m)
    {
        var lines = new List<string> { $"Lecture: {m.Title}", $"Date: {Py.Head(m.Date, 10)}" };
        if (m.Folder.Length > 0) lines.Add($"Granola folder: {m.Folder}");
        return string.Join("\n", lines);
    }

    public static string WholePrompt(Meeting m, string transcript) =>
        "You are an expert note-taker for university lectures. Write study notes for this lecture "
        + $"from its transcript.\n\n{Structure}\n\n{Rules}\n\n{Header(m)}\n\nTranscript:\n{transcript}";

    public static string PartPrompt(Meeting m, string part, int i, int n) =>
        $"You are taking notes on part {i} of {n} of a lecture transcript. Write detailed Markdown bullet "
        + "notes for this part only: every concept, definition, example, formula, and announcement, in order. "
        + "They will be merged with the other parts later, so write no introduction or conclusion.\n\n"
        + $"{Rules}\n\n{Header(m)}\n\nTranscript part {i} of {n}:\n{part}";

    public static string CondensePrompt(Meeting m, IReadOnlyList<string> notes)
    {
        string joined = string.Join("\n\n", notes.Select((n, i) => $"### Notes {i + 1}\n{n}"));
        return "Combine these notes on consecutive parts of one lecture into one set of detailed Markdown bullet "
            + $"notes. Remove repetition but keep every distinct fact.\n\n{Rules}\n\n{Header(m)}\n\n{joined}";
    }

    public static string MergePrompt(Meeting m, IReadOnlyList<string> notes)
    {
        string joined = string.Join("\n\n", notes.Select((n, i) => $"### Part {i + 1}\n{n}"));
        return "You are an expert note-taker for university lectures. Below are notes on consecutive parts of one "
            + "lecture. Merge them into one set of study notes, removing repetition and keeping every distinct "
            + $"fact.\n\n{Structure}\n\n{Rules}\n\n{Header(m)}\n\n{joined}";
    }

    /// <summary>Our own study notes for a lecture, written from its transcript.</summary>
    public static async Task<string> SummarizeTranscriptAsync(Meeting m, Config cfg, ChatFn? chat = null, ShowFn? show = null)
    {
        chat ??= (c, model, prompt, ctx) => OllamaGenerateAsync(c, model, prompt, ctx);
        string model = cfg.EffectiveSummaryModel;
        int ctx = await ContextSizeAsync(cfg, model, show);
        int budget = TranscriptBudget(ctx);
        string text = Py.Strip(TimedText.Plain(m.Transcript)); // a recording's times would only distract the model
        if (text.Length <= budget) return CleanOutput(await chat(cfg, model, WholePrompt(m, text), ctx));
        var parts = SplitTranscript(text, budget);
        var notes = new List<string>();
        for (int i = 0; i < parts.Count; i++)
            notes.Add(CleanOutput(await chat(cfg, model, PartPrompt(m, parts[i], i + 1, parts.Count), ctx)));
        // Very long lectures: fold pairs of part-notes together until the merge fits in one call.
        while (notes.Count > 1 && notes.Sum(n => n.Length) > budget)
        {
            var folded = new List<string>();
            for (int i = 0; i < notes.Count; i += 2)
                folded.Add(i + 1 < notes.Count ? CleanOutput(await chat(cfg, model, CondensePrompt(m, notes[i..(i + 2)]), ctx)) : notes[i]);
            notes = folded;
        }
        return CleanOutput(await chat(cfg, model, MergePrompt(m, notes), ctx));
    }
}
